pep0/pep.py

```diff
@@ -109,28 +113,64 @@ def sort_by(self):
             base = self.last.lower()
         return unicodedata.normalize('NFKD', base).encode('ASCII', 'ignore')
 
-    def _last_name(self, full_name):
-        """Find the last name (or nickname) of a full name.
+    @staticmethod
+    def _parse_name(full_name):
+        """Decompose a full name into parts.
 
-        If no last name (e.g, 'Aahz') then return the full name.  If there is
-        a leading, lowercase portion to the last name (e.g., 'van' or 'von')
-        then include it.  If there is a suffix (e.g., 'Jr.') that is appended
-        through a comma, then drop the suffix.
+        If a mononym (e.g, 'Aahz') then return the full name. If there are
+        suffixes in the name (e.g. ', Jr.' or 'III'), then find and extract
+        them. If there is a middle initial followed by a full stop, then
+        combine the following words into a surname (e.g. N. Vander Weele). If
+        there is a leading, lowercase portion to the last name (e.g. 'van' or
+        'von') then include it in the surname.
 
         """
-        name_partition = full_name.partition(u',')
-        no_suffix = name_partition[0].strip()
-        suffix = name_partition[2].strip()
-        name_parts = no_suffix.split()
-        part_count = len(name_parts)
-        if part_count == 1 or part_count == 2:
-            return name_parts[-1], suffix
-        else:
-            assert part_count > 2
+        possible_suffixes = ["Jr", "Jr.", "II", "III"]
+        special_cases = ["The Python core team and community"]
+
+        if full_name in special_cases:
+            return {"name": full_name}
+
+        suffix_partition = full_name.partition(",")
+        pre_suffix = suffix_partition[0].strip()
+        suffix = suffix_partition[2].strip()
+
+        name_parts = pre_suffix.split(" ")
+        num_parts = len(name_parts)
+        name = {"suffix": suffix}
+
+        if num_parts == 0:
+            raise ValueError("Name is empty!")
+        elif num_parts == 1:
+            name.update(name=name_parts[0])
+        elif num_parts == 2:
+            name.update(forename=name_parts[0], surname=name_parts[1])
+        elif num_parts > 2:
+            # handles III etc.
+            if name_parts[-1] in possible_suffixes:
+                new_suffix = " ".join([*name_parts[-1:], suffix]).strip()
+                name_parts.pop(-1)
+                name.update(suffix=new_suffix)
+
+            # handles von, van, v. etc.
             if name_parts[-2].islower():
-                return u' '.join(name_parts[-2:]), suffix
+                forename = " ".join(name_parts[:-2])
+                surname = " ".join(name_parts[-2:])
+                name.update(forename=forename, surname=surname)
+
+            # handles double surnames after a middle initial (e.g. N. Vander Weele)
```

Comment:
This is tough, a name like `R. David Murray` should give `Murray` as a surname for the PEP index.

That said, a name written in Japanese English convention like `INADA Naoki` should return `INADA` as surname.
In short, it’s not generally possible to parse world names into US categories of «first», «middle», «last»

I guess we have to accept imperfect for now, add special cases when we notice problems (what PEP will add the first `Name MacName Sr` :slightly_smiling_face:), and maybe someday rework the system to have the proper way: metadata (or some author index dict) should include full name and short name for PEP 0.

Reply:
Luckily (?!) in PEP 458 `R. David Murray` is listed without the full stop in his first initial, so the name is still parsed correctly. I don't think there's a good solution for `INADA Naoki` beyond adding a special-case exception

> In short, it’s not generally possible to parse world names into US categories of «first», «middle», «last»

Or UK, Australia, Canada etc :stuck_out_tongue_winking_eye:. But I get your point. I'm reminded of [this post](https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/) about names - hopefully #&#8203;40 doesn't apply to us...

I think that your last suggestion having some sort of lookup table is probably the best solution, as in all the PEPs there are still only a relativley small number of authors (248) - it's quite late here so will add that feature tommorow. It also keeps special cases etc. out of the code to keep it from becoming knobbly.

Reply:
It would be OK if this PEP fixed the most egregious cases (`III`) without covering 100% of possibilities :slightly_smiling_face:

Reply:
Gives me something to do! Latest commit adds such a metadata lookup and therefore simplifies the name parsing code. 

This should make it so that names can be correctly entered into AUTHORS.csv and PEP 0 will reflect this. I've also identified some duplicate entries (e.g. `P.J. Eby` &  `Phillip J. Eby`, `Greg` and `Gregory` Ewing, `Jim J. Jewett` & Jim Jewett`, `Martin v. Löwis` & `Martin von Löwis`). Is it acceptable to modify PEP headers to canonicalise these names?

Reply:
I think I would add multiple entries to the data file rather than editing historical documents.

Reply:
The latest data file (exceptions rather than full mapping) doesn’t de-duplicate these entries, should it?

Reply:
The data file is checked first (in __init__), so unsure where duplicates would propogate from? 

Always good to be preventative but not sure I understand this one, sorry!

Reply:
Maybe my comment doesn’t make sense!
I don’t have a clear picture of the current behaviour of the code, so I wondered if the change from full data file to exceptions data file did preserve the feature you added of normalizing the duplicate names.

Reply:
Aha! I forgot to add that back in, you're right - have done so now (only adding the less used variant and mapping it to the 'cannonical' variant, to keep the file smaller)